theano/gpuarray/dnn.py

```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    """
+    This Op builds a spatial transformer descriptor for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dimensions', 'dtype')
+    params_type = ParamsType(nimages=int_t, nchannels=int_t, height=int_t, width=int_t,
+                             dtype=cudnn.cudnnDataType_t)
+
+    def c_headers(self):
+        return ['cudnn.h', 'cudnn_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), config.dnn.include_path]
+
+    def c_libraries(self):
+        return ['cudnn']
+
+    def c_lib_dirs(self):
+        return [config.dnn.library_path]
+
+    def do_constant_folding(self, node):
+        return False
+
+    def __init__(self, dimensions, dtype="float32"):
+        COp.__init__(self, ["c_code/spatialtf_desc.c"], "APPLY_SPECIFIC(spatialtf_desc)")
+
+        # cuDNN supports only 2D transformations, therefore output tensor must
+        # have exactly 4 dimensions: (width, height, num_channels, num_images)
+        assert len(dimensions) == 4
+        self.dimensions = tuple(dimensions)
+
+        assert cudnn.cudnnDataType_t.has_alias(dtype)
+        self.dtype = dtype
+
+    def make_node(self):
+        node = Apply(self, [],
+                     [CDataType("cudnnSpatialTransformerDescriptor_t",
+                                freefunc="cudnnDestroySpatialTransformerDescriptor")()])
+        # DebugMode cannot compare the values of CDataType variables, so by
+        # default it returns False all the time. To prevent DebugMode from
+        # complaining because of the MergeOptimizer, we make this variable
+        # always compare to True.
+        out = node.outputs[0]
+        out.tag.values_eq_approx = tensor.type.values_eq_approx_always_true
+        return node
+
+    # Number of images
+    nimages = property(lambda self: self.dimensions[0])
+    # Number of channels
+    nchannels = property(lambda self: self.dimensions[1])
+    # Grid height
+    height = property(lambda self: self.dimensions[2])
+    # Grid width
+    width = property(lambda self: self.dimensions[3])
+
+    def c_code_cache_version(self):
+        return (super(GpuDnnSpatialTfDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnGridGenerator(DnnBase):
+
+    """
+    This Op builds a spatial transformer grid generator for use in spatial transformer network
+    operations.
+    """
+
+    __props__ = ('dtype',)
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+
+    def __init__(self, dtype):
+        DnnBase.__init__(self, ["c_code/spatialtf_grid.c"], "spatialtf_grid")
+        self.dtype = dtype
+
+    def make_node(self, grid_dimensions, theta, desc):
+        context_name = infer_context_name(desc, theta)
+
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert theta.dtype in ('float16', 'float32', 'float64')
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
```

Comment:
We can remove this line, as previous line already checks `theta.dtype` with a fixed set of supported types.